### ISSUE TITLE
EoT is not accounting for transformations in backward pass

### ISSUE THREAD

**Issue by mathsinn** [Trusted-AI/adversarial-robustness-toolbox#289]
**Describe the bug**
Currently, the loss and class gradient calculations for EoT account for transformations only in the forward pass. This is inconsistent with the definition e.g. in https://arxiv.org/pdf/1802.00420.pdf, Section 4.2.

**To Reproduce**
This is revealed by a static code analysis: l.90 and l.108 in https://github.com/IBM/adversarial-robustness-toolbox/blob/master/art/wrappers/expectation.py

**Expected behavior**
Transformations ought to be accounted for in the backward pass.

**Screenshots**
N/A

**System information (please complete the following information):**
This is a system-independent bug in the implementation of EoT

**Comment by beat-buesser** [comment]
Hi @mathsinn Thank you very much for opening this issue!

**Comment by Embeddave** [comment]
hi @beat-buesser @mathsinn are you still considering this a bug?
I don't see any related commits to the code since this issue was opened.

I admit it's not clear to me why the implementation wouldn't work as is.
Why would you need to change anything about the backward pass?
The transformations are already accounted for, in the sense that eager mode includes those computations in the graph.

It is true that BPDA as defined in section 4.1.2 of the paper @mathsinn cites propose to replace a layer f(x) with another layer on the backward pass. Is that what you're referring to?

But I don't think that's related to EOT, which as I understand it is just math justifying why you can still generate adversarial examples using gradient descent, even if the defender applies random transforms. Happy to hear what I'm getting wrong tho

**Comment by Embeddave** [comment]
I did just try a Carlini-Wagner L2 attack against a classifier wrapped in EoT, and none of the attacks were successful ...
but it's really not clear to me if I'm defining the iterator the right way. 
I'm trying to follow the example from the tests that uses FGSM

**Comment by mathsinn** [comment]
Hi @Embeddave 
If you look e.g. at this line in the loss gradient function: https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/wrappers/expectation.py#L98
What the issue refers to is that we aren't explicitly taking into account the gradients of the transformation function. Instead, implicitly, we apply BPDA with the straight-through estimator. This will work reasonably well in cases where the transformation functions can be approximated by the identity function; also it supports taking expectations over classifiers' inherent randomness, e.g. if drop-out is used at inference time. In general however this approximation may be too coarse and is a deficiency that we shall address in a future release of ART.

Re your specific use case where EoT attacks weren't successful: could you share more details on how you are using the iterator?

**Comment by Embeddave** [comment]
thank you @mathsinn that helps.

I think I see now why EOT needs to account for the transformation in the backward pass: in practice we take the expectation of the gradient, so we need a transform that's differentiable (or an approximation like in BPDA). 
It's clearer to me from the Athalye 2081 notation (like in section 4.2) than from "Synthezing Robust Adversarial Examples"

Here's a code snippet of how I'm using it -- please let me know if I'm not setting things up right
Copied from this test
https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/307f7c9ec255c9c3594a9e57c82a2b9064c69843/tests/wrappers/test_expectation.py#L68
Sorry this is so verbose, trying to give you any info about set-up that I might be getting wrong -- mainly not sure if I define the iterator correctly
```python
from collections import defaultdict
import functools

from art.attacks.evasion.carlini import CarliniL2Method
from art.estimators.classification import PyTorchClassifier
from art.wrappers.expectation import ExpectationOverTransformations
import numpy as np
from torch import nn
from torch.optim import SGD
from tqdm import tqdm

resnet = resnet50()  # pre-trained from Madry lab `robustness` library
test_loader = get_loader()  # from `robustness`
normalizer = get_normalizer()  # again from `robustness`

optimizer = SGD(resnet.parameters())
criterion = nn.CrossEntropyLoss()
mean = np.asarray(CIFAR10_MEAN).reshape((3, 1, 1))
std = np.asarray(CIFAR10_STD).reshape((3, 1, 1))
clf_art = PyTorchClassifier(
    model=resnet,
    clip_values=(0., 1.),
    loss=criterion,
    optimizer=optimizer,
    input_shape=(3, 32, 32),
    nb_classes=10,
    preprocessing=(mean, std)
)

BATCH_SIZE = 4
cw_attack = CarliniL2Method(classifier=clf_art, batch_size=BATCH_SIZE)

PARAM = 0.25
# pre-processing defense transform is a distant cousin of total variance minimization
# we wrap it in functools.partial so we can set params
_tv_like = functools.partial(our_package.tv_like, param=PARAM)

def transformation():
    yield _tv_like

eot = ExpectationOverTransformations(classifier=clf_art, sample_size=1, transformation=transformation)
cw_eot = CarliniL2Method(classifier=eot, batch_size=BATCH_SIZE)

DEVICE = 'cuda'
correct = defaultdict(list)
for batch_num, batch in enumerate(tqdm(test_loader)):
    if batch_num > N_TEST_BATCHES:
        break
    
    # measure unattacked accuracy
    x, y = batch[0].to(DEVICE), batch[1].to(DEVICE)
    with torch.no_grad():
        y_pred = resnet(normalizer(x)).argmax(dim=1)
    correct['x'].extend(
        (y_pred == y).cpu().numpy().tolist()
    )

    # attack tv_like with Carlini Wagner using Expectation over Transformation
    x_cw_eot = cw_eot.generate(x)
    x_cw_eot_tensor = torch.from_numpy(x_cw_eot).to(DEVICE)
    with torch.no_grad():
        y_pred = resnet(normalizer(x_cw_eot_tensor)).argmax(dim=1)
    correct['x_cw_eot'].extend(
        (y_pred == y).cpu().numpy().tolist()
    )
```

which gives me
```
for k, v in correct.items():
    print(f'{k} : num correct: {sum(v)}, percent: {(sum(v) / len(v)) * 100}')

x : num correct: 16, percent: 100.0
x_cw_eot : num correct: 16, percent: 100.0
```
... which is surprising since I know CW attack does have some success rate on our defense when I am *not* using the EOT wrapper

**Comment by mathsinn** [comment]
@Embeddave Thank you for sharing! Can you say a bit more about the type of transformations in `our_package.tv_like`? In particular:
- are those transformations stochastic? I.e. will the iterator always return the same function, or is there some inherent randomness?
- I suppose those transformations will map inputs onto the same domain (3x32x32)? Do they roughly act like the identity (e.g. like JPEG compression preprocessing they slightly alter the input, but the identify function is a good approximation)?

**Comment by Embeddave** [comment]
> are those transformations stochastic? I.e. will the iterator always return the same function, or is there some inherent randomness?

the function is the same but it performs a reconstruction in a stochastic manner

> Do they roughly act like the identity (e.g. like JPEG compression preprocessing they slightly alter the input, but the identify function is a good approximation)?

whether or not g(x) ~ x is debatable, but I think it's fair to say in our case the identity function is a __less__ good approximation than it is to TVM

so, an attacker would need to approximate our defense w/something differentiable.
The problem is that even for something where g(x)~x holds, like TVM, I don't find any implementations, e.g. in the `obfuscated-gradients` repo, where they meaningfully perform EOT. Instead it appears to be the case that people just take a "mean with a sample size of 1", i.e. they just apply gradient descent to a single image that's transformed once for every step. Would love to find out I'm wrong so I can better understand how their attack works. An exception is patch attack but of course there we've given up on staying within some epsilon ball.

**Comment by mathsinn** [comment]
Ok, since your function has a stochastic element, a `sample_size>1` may be needed to generate adversarial examples that transfer across the different reconstructions. The minimum appropriate value for `sample_size` depends on the distribution of the reconstructions. Maybe start with a `sample_size` as large as you can computationally afford on a small batch, see whether it works, and if yes then you could try reducing it.

**Comment by Embeddave** [comment]
will do, thank you @mathsinn

**Comment by beat-buesser** [comment]
Expectation over Transformation operations accounting for transformations/gradients in backward passes should now be supported with `art.preprocessing.expectation_over_transformation` starting from ART 1.5.0.
